sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/redactionExpressions.scala

```diff
@@ -0,0 +1,237 @@
+      > SELECT _FUNC_("1234567898765432");
+        NULL
+      > SELECT _FUNC_("1234567898765432", "XXXX-XXXX-XXXX-1234");
+        NULL
+  """,
+  since = "3.4.0",
+  group = "string_funcs"
+)
+case class TryMaskCcn(left: Expression, right: Expression)
+  extends MaskDigitSequence(left, right, "try_mask_ccn", true) {
+  def this(left: Expression) = this(left, Literal(MaskCcn.DEFAULT_FORMAT))
+  override protected def withNewChildrenInternal(
+      newInput: Expression, newFormat: Expression): TryMaskCcn =
+    copy(left = newInput, right = newFormat)
+  override def nullable: Boolean = true
+}
+
+/** Companion object for the MaskCcn and TryMaskCcn classes. */
+object MaskCcn {
+  val DEFAULT_FORMAT = "XXXX-XXXX-XXXX-XXXX"
+}
+
+/** Implementation of an expression to mask digits in a string to replacement characters. */
+abstract class MaskDigitSequence(
+    left: Expression, right: Expression, functionName: String, nullOnError: Boolean)
+  extends BinaryExpression with ImplicitCastInputTypes with NullIntolerant with Serializable {
+  private def format: Expression = right
+
+  private lazy val formatString = if (format.foldable) {
+    format.eval().toString.toUpperCase(Locale.ROOT)
+  } else {
+    ""
+  }
+  private lazy val parser = new MaskDigitSequenceParser(formatString, nullOnError)
+
+  override def prettyName: String = functionName
+  override def dataType: DataType = StringType
+  override def inputTypes: Seq[DataType] = Seq(StringType, StringType)
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val inputTypeCheck = super.checkInputDataTypes()
+    if (inputTypeCheck.isSuccess) {
+      val formatStringValid = formatString.forall {
+        case 'x' | 'X' | '-' => true
+        case ch if ch.isDigit || ch.isWhitespace => true
+        case _ => false
+      }
+      if (formatStringValid) {
+        TypeCheckResult.TypeCheckSuccess
+      } else {
+        throw QueryCompilationErrors.maskCcnInvalidFormatError(prettyName)
+      }
+    } else {
+      inputTypeCheck
+    }
+  }
+
+  override def nullSafeEval(string: Any, format: Any): Any =
+    parser.parse(string.asInstanceOf[UTF8String])
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val builder = ctx.addReferenceObj(
+      "parser", parser, classOf[MaskDigitSequenceParser].getName)
+    val eval = left.genCode(ctx)
+    ev.copy(code =
+      code"""
+        |${eval.code}
+        |boolean ${ev.isNull} = ${eval.isNull};
+        |${CodeGenerator.javaType(dataType)} ${ev.value} = ${CodeGenerator.defaultValue(dataType)};
+        |if (!${ev.isNull}) {
+        |  ${ev.value} = $builder.parse(${eval.value});
+        |  if (${ev.value} == null) {
+        |    ${ev.isNull} = true;
+        |  }
+        |}
+      """.stripMargin)
+  }
+}
+
+/** Executes the string parsing steps for the MaskDigitSequence class. */
+class MaskDigitSequenceParser(formatString: String, nullOnError: Boolean) extends Serializable {
```

Comment:
why extending Serializable

Reply:
This one was not necessary, I have removed it.